README
NAME
    Return::Type - specify a return type for a function (optionally with
    coercion)

SYNOPSIS
       use Return::Type;
       use Types::Standard qw(Int);
   
       sub first_item :ReturnType(Int) {
          return $_[0];
       }
   
       my $answer = first_item(42, 43, 44);     # returns 42
       my $pie    = first_item(3.141592);       # throws an error!

DESCRIPTION
    Return::Type allows you to specify a return type for your subs. Type
    constraints from any Type::Tiny, MooseX::Types or MouseX::Types type
    library are supported.

    The simple syntax for specifying a type constraint is shown in the
    "SYNOPSIS". If the attribute is passed a single type constraint as shown,
    this will be applied to the return value if called in scalar context, and
    to each item in the returned list if called in list context. (If the sub
    is called in void context, type constraints are simply ignored.)

    It is possible to specify different type constraints for scalar and list
    context:

       sub foo :ReturnType(scalar => Int, list => HashRef[Num]) {
          if (wantarray) {
             return (pie => 3.141592);
          }
          else {
             return 42;
          }
       }

    Note that because type constraint libraries are really aimed at validating
    scalars, the type constraint for the list is specified as a *hashref* of
    numbers and not a hash of numbers! For the purposes of validation against
    the type constraint, we slurp the returned list into a temporary arrayref
    or hashref.

    For type constraints with coercions, you can also pass the option `coerce
    => 1`:

       use Return::Type;
       use Types::Standard qw( Int Num );
   
       # Define a subtype of "Int" at compile time, which can
       # coerce from "Num" by rounding to nearest integer.
       use constant Rounded => Int->plus_coercions(Num, sub { int($_) });
   
       sub first_item :ReturnType(scalar => Rounded, coerce => 1) {
          return $_[0];
       }
   
       my $answer = first_item(42, 43, 44);     # returns 42
       my $pie    = first_item(3.141592);       # returns 3

    The options `coerce_scalar` and `coerce_list` are also available if you
    wish to enable coercion only in particular contexts.

  Power-user Inferface
    Rather than using the `:ReturnType` attribute, it's possible to wrap a
    coderef like this:

       my $wrapped = Return::Type->wrap_sub($orig, %options);

    The accepted options are `scalar`, `list`, `coerce`, `coerce_list`, and
    `coerce_scalar`, as per the attribute-based interface.

    There is an additional option `scope_upper` which will load and use
    Scope::Upper so that things like `caller` used within the wrapped sub are
    unaware of being wrapped. This behaviour was the default prior to
    Return::Type 0.004, but is now optional and disabled by default.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=Return-Type>.

SUPPORT
    IRC: support is available through in the *#moops* channel on irc.perl.org
    <http://www.irc.perl.org/channels.html>.

SEE ALSO
    Attribute::Contract, Sub::Filter, Sub::Contract.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2013 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Return-Type
===========

Created:      2013-10-30
Home page:    <https://metacpan.org/release/Return-Type>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Return-Type>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.004	2014-07-30

 [ Packaging ]
 - No longer require Scope::Upper.

 [ Other ]
 - Make the Scope::Upper stuff optional, and disabled by default.

0.003	2013-11-14

 [ Documentation ]
 - Fix typo in SYNOPSIS code.

0.002	2013-11-03

 [ Bug Fixes ]
 - Fix non-inlineble type constraints and coercions.

 [ Other ]
 - Allow per-context (i.e. different for scalar and list contexts) control
   over coercion.

0.001	2013-10-30	Initial release


lib/Return/Type.pm
use 5.008;
use strict;
use warnings;

package Return::Type;

our $AUTHORITY = 'cpan:TOBYINK';
our $VERSION   = '0.001';

use Attribute::Handlers;
use Eval::TypeTiny qw( eval_closure );
use Scope::Upper qw( );
use Sub::Identify qw( sub_fullname );
use Sub::Name qw( subname );
use Types::Standard qw( Any ArrayRef HashRef Int );
use Types::TypeTiny qw( to_TypeTiny );

sub _inline_type_check
{
	my $class = shift;
	my ($type, $var, $env) = @_;
	
	return $type->inline_assert($var) if $type->can_be_inlined;
	
	$env->{'$type'} = \$type;
	return sprintf('$type->assert_return(%s);', $var);
}

sub _inline_type_coerce_and_check
{
	my $class = shift;
	my ($type, $var, $env) = @_;
	
	my $coerce = '';
	if ($type->has_coercion and $type->coercion->can_be_inlined)
	{
		$coerce = sprintf('%s = %s;', $var, $type->coercion->inline_coercion($var));
	}
	elsif ($type->has_coercion)
	{
		$env->{'$coercion'} = \( $type->coercion );
		$coerce = sprintf('%s = $coercion->coerce(%s);', $var, $var);
	}
	
	return $coerce . $class->_inline_type_check(@_);
}

sub wrap_sub
{
	my $class = shift;
	my $sub   = $_[0];
	local %_  = @_[ 1 .. $#_ ];
	
	$_{$_}     &&= to_TypeTiny($_{$_}) for qw( list scalar );
	$_{scalar} ||= Any;
	$_{list}   ||= ($_{scalar} == Any ? Any : ArrayRef[$_{scalar}]);
	
	my $prototype = prototype($sub);
	$prototype = defined($prototype) ? "($prototype)" : "";
	
	my %env  = ( '$sub' => \$sub );
	my @src  = sprintf('sub %s { my $wa = wantarray;', $prototype);
	my $call = '&Scope::Upper::uplevel($sub => (@_) => &Scope::Upper::SUB(&Scope::Upper::SUB))';
	
	my $inline = $_{coerce} ? '_inline_type_coerce_and_check' : '_inline_type_check';
	
	# List context
	push @src, 'if ($wa) {';
	if ( $_{list}->is_a_type_of(HashRef) )
	{
		push @src, 'my $rv = do { use warnings FATAL => qw(misc); +{' . $call . '} };';
		push @src, $class->$inline($_{list}, '$rv', \%env);
		push @src, 'return %$rv;';
	}
	else
	{
		push @src, 'my $rv = [' . $call . '];';
		push @src, $class->$inline($_{list}, '$rv', \%env);
		push @src, 'return @$rv;';
	}
	push @src, '}';
	
	# Scalar context
	push @src, 'elsif (defined $wa) {';
	push @src, 'my $rv = ' . $call . ';';
	push @src, $class->$inline($_{scalar}, '$rv', \%env);
	push @src, 'return $rv;';
	push @src, '}';
	
	# Void context - cannot request a value to check, so check must be skipped
	push @src, 'goto $sub;';
	
	push @src, '}';
	
	my $rv = eval_closure(
		source       => \@src,
		environment  => \%env,
	);
	return subname(sub_fullname($sub), $rv);
}

sub UNIVERSAL::ReturnType :ATTR(CODE)
{
	my ($package, $symbol, $referent, $attr, $data) = @_;
	
	no warnings qw(redefine);
	my %args = (@$data % 2) ? (scalar => @$data) : @$data;
	*$symbol = __PACKAGE__->wrap_sub($referent, %args);
}

1;

__END__

=pod

=encoding utf-8

=head1 NAME

Return::Type - specify a return type for a function (optionally with coercion)

=head1 SYNOPSIS

   use Return::Type;
   use Types::Standard qw(Int);
   
   sub first_item :ReturnType(Int) {
      return $_[0];
   }
   
   my $answer = first_item(42, 43, 44);     # returns 44
   my $pie    = first_item(3.141592);       # throws an error!

=head1 DESCRIPTION

Return::Type allows you to specify a return type for your subs. Type
constraints from any L<Type::Tiny>, L<MooseX::Types> or L<MouseX::Types>
type library are supported.

The simple syntax for specifying a type constraint is shown in the
L</SYNOPSIS>. If the attribute is passed a single type constraint as shown,
this will be applied to the return value if called in scalar context, and
to each item in the returned list if called in list context. (If the sub
is called in void context, type constraints are simply ignored.)

It is possible to specify different type constraints for scalar and
list context:

   sub foo :ReturnType(scalar => Int, list => HashRef[Num]) {
      if (wantarray) {
         return (pie => 3.141592);
      }
      else {
         return 42;
      }
   }

Note that because type constraint libraries are really aimed at
validating scalars, the type constraint for the list is specified as
a I<hashref> of numbers and not a hash of numbers! For the purposes
of validation against the type constraint, we slurp the returned list
into a temporary arrayref or hashref.

For type constraints with coercions, you can also pass the option
C<< coerce => 1 >>:

   use Return::Type;
   use Types::Standard qw( Int Num );
   
   # Define a subtype of "Int" at compile time, which can
   # coerce from "Num" by rounding to nearest integer.
   use constant Rounded => Int->plus_coercions(Num, sub { int($_) });
   
   sub first_item :ReturnType(scalar => Rounded, coerce => 1) {
      return $_[0];
   }
   
   my $answer = first_item(42, 43, 44);     # returns 44
   my $pie    = first_item(3.141592);       # returns 3

=head2 Power-user Inferface

Rather than using the C<< :ReturnType >> attribute, it's possible to
wrap a coderef like this:

   my $wrapped = Return::Type->wrap_sub($orig, %options);

The accepted options are C<scalar>, C<list> and C<coerce>, as per the
attribute-based interface.

=begin trustme

=item wrap_sub

=end trustme

=head1 BUGS

Please report any bugs to
L<http://rt.cpan.org/Dist/Display.html?Queue=Return-Type>.

=head1 SEE ALSO

L<Attribute::Contract>,
L<Sub::Filter>,
L<Sub::Contract>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT AND LICENCE

This software is copyright (c) 2013 by Toby Inkster.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 DISCLAIMER OF WARRANTIES

THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "specify a return type for a function (optionally with coercion)",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>", "unknown"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.016, CPAN::Meta::Converter version 2.140640",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "Return-Type",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { recommends => { "Dist::Inkt" => 0.001 } },
                        runtime   => {
                                       requires => {
                                                     "Eval::TypeTiny" => 0,
                                                     "perl" => 5.008,
                                                     "Sub::Identify" => 0,
                                                     "Sub::Name" => 0,
                                                     "Types::Standard" => 0,
                                                     "Types::TypeTiny" => 0,
                                                   },
                                       suggests => { "Scope::Upper" => 0 },
                                     },
                        test      => { requires => { "Test::Fatal" => 0, "Test::More" => 0.96 } },
                      },
  "provides"       => {
                        "Return::Type" => { file => "lib/Return/Type.pm", version => 0.004 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => { web => "http://rt.cpan.org/Dist/Display.html?Queue=Return-Type" },
                        homepage     => "https://metacpan.org/release/Return-Type",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          url  => "git://github.com/tobyink/p5-return-type.git",
                                          web  => "https://github.com/tobyink/p5-return-type",
                                        },
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/Return-Type/project",
                      },
  "version"        => 0.004,
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

